pandas/tests/typing/valid/test_frame.py

```diff
@@ -0,0 +1,743 @@
+import io
+from pathlib import Path
+import tempfile
+from typing import (
+    Any,
+    Dict,
+    Iterable,
+    List,
+    Tuple,
+)
+
+import numpy as np
+
+import pandas as pd
+
+
+def test_types_init() -> None:
+    pd.DataFrame(data={"col1": [1, 2], "col2": [3, 4]})
+    pd.DataFrame(data={"col1": [1, 2], "col2": [3, 4]}, index=[2, 1])
+    pd.DataFrame(data=[1, 2, 3, 4], dtype=np.int8)
+    pd.DataFrame(
+        np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]]),
+        columns=["a", "b", "c"],
+        dtype=np.int8,
+        copy=True,
+    )
+
+
+def test_types_to_csv() -> None:
+    df = pd.DataFrame(data={"col1": [1, 2], "col2": [3, 4]})
+    csv_df: str = df.to_csv()
+
+    with tempfile.NamedTemporaryFile() as file:
+        df.to_csv(file.name)
+        df2: pd.DataFrame = pd.read_csv(file.name)
+
+    with tempfile.NamedTemporaryFile() as file:
+        df.to_csv(Path(file.name))
+        df3: pd.DataFrame = pd.read_csv(Path(file.name))
+
+    # This keyword was added in 1.1.0 https://pandas.pydata.org/docs/whatsnew/v1.1.0.html
+    with tempfile.NamedTemporaryFile() as file:
+        df.to_csv(file.name, errors="replace")
+        df4: pd.DataFrame = pd.read_csv(file.name)
+
+    # Testing support for binary file handles, added in 1.2.0 https://pandas.pydata.org/docs/whatsnew/v1.2.0.html
+    df.to_csv(io.BytesIO(), encoding="utf-8", compression="gzip")
+
+
+def test_types_to_csv_when_path_passed() -> None:
+    df = pd.DataFrame(data={"col1": [1, 2], "col2": [3, 4]})
+    path: Path = Path("./dummy_path.txt")
+    try:
+        assert not path.exists()
+        df.to_csv(path)
+        df5: pd.DataFrame = pd.read_csv(path)
+    finally:
+        path.unlink()
+
+
+def test_types_copy() -> None:
+    df = pd.DataFrame([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
+    df2: pd.DataFrame = df.copy()
+
+
+def test_types_getitem() -> None:
+    df = pd.DataFrame(data={"col1": [1, 2], "col2": [3, 4], 5: [6, 7]})
+    i = pd.Index(["col1", "col2"])
+    s = pd.Series(["col1", "col2"])
+    select_df = pd.DataFrame({"col1": [True, True], "col2": [False, True]})
+    a = np.array(["col1", "col2"])
+    df["col1"]
+    df[5]
+    df[["col1", "col2"]]
+    df[1:]
+    df[s]
+    df[a]
+    df[select_df]
+    df[i]
+
+
+def test_slice_setitem() -> None:
+    # Due to the bug in pandas 1.2.3(https://github.com/pandas-dev/pandas/issues/40440), this is in separate test case
+    df = pd.DataFrame(data={"col1": [1, 2], "col2": [3, 4], 5: [6, 7]})
+    df[1:] = ["a", "b", "c"]
+
+
+def test_types_setitem() -> None:
+    df = pd.DataFrame(data={"col1": [1, 2], "col2": [3, 4], 5: [6, 7]})
+    i = pd.Index(["col1", "col2"])
+    s = pd.Series(["col1", "col2"])
+    select_df = pd.DataFrame({"col1": [True, True], "col2": [False, True]})
+    a = np.array(["col1", "col2"])
+    df["col1"] = [1, 2]
+    df[5] = [5, 6]
+    df[["col1", "col2"]] = [[1, 2], [3, 4]]
+    df[s] = [5, 6]
+    df[a] = [[1, 2], [3, 4]]
+    df[select_df] = [1, 2, 3]
```

Comment:
This does not work: `ValueError: cannot assign mismatch length to masked array`

Reply:
Yup. Still looking at this. Investigating this as a possible regression from pandas version `1.3.2`.

Reply:
Type tests finding non-typing bugs :) If this worked in previous versions, feel free to open an issue!

Reply:
That's not first time that this happened. See https://github.com/pandas-dev/pandas/issues/40440 originally spotted by @joannasendorek.

I've opened up an issue: https://github.com/pandas-dev/pandas/issues/45593. I'll circumvent this for now in my PR which is pending.